docs: introduce badging levels (#242)
defenseunicorns/uds-common: docs/package_integration/uds-package-practices.md

# UDS Package Practices

> This document describes the standards for [<img alt="Made for UDS" src="../made-for-uds.svg" height="20px"/>](https://github.com/defenseunicorns/uds-core) badging. It is not a comprehensive guide to creating UDS Packages, a more detailed guide can be found [here](guide.md).

Author: updated

> [!NOTE]

> > This document assumes knowledge of the UDS ecosystem adn the UDS Package Custom Resource. If you are unfamiliar with these concepts, please first refer to the [guide](guide.md) providing more detailed information.

Reviewer: ```suggestion
> This document assumes knowledge of the UDS ecosystem and the UDS Package Custom Resource. If you are unfamiliar with these concepts, please first refer to the [guide](guide.md) providing more detailed information.
```

This document outlines services and features of [UDS Core](https://github.com/defenseunicorns/uds-core) UDS package's **must** integrate with, using the [UDS `Package` custom resource](https://github.com/defenseunicorns/uds-core/blob/main/src/pepr/operator/README.md#example-uds-package-cr).

Made for UDS packages can be one of three tiers:

[<img alt="Gold" src="../made-for-uds-gold.svg" height="20px"/>](https://github.com/defenseunicorns/uds-core), [<img alt="Silver" src="../made-for-uds-silver.svg" height="20px"/>](https://github.com/defenseunicorns/uds-core), or [<img alt="bronze" src="../made-for-uds-bronze.svg" height="20px"/>](https://github.com/defenseunicorns/uds-core).

> [!IMPORTANT]

> > Packages should aim for Gold by default and only _SETTLE_ for lesser tiers of Bronze and Silver.

Reviewer 2: ```suggestion
> Packages should aim for Gold by default and only _SETTLE_ for lesser tiers of Bronze and Silver by documented exception.
```

is it worth calling out or linking to examples of documented exceptions here like we did in the bounty program directly? those examples will continue being applicable after the program ends

## Gold

_a Gold UDS-Package out of the box implements best-effort 0-cve images, configuration hardening, and meet’s the unicorn guarantee with zero additional effort._

Gold Packages:

- **Must** Satisfy all the requirements of [Silver](#silver) packages
- **Must** include OSCAL-component control mapping and responses for the application. see [OSCAL Guidelines](oscal-guidelines.md)
- **Must** minimize the scope and number of the exemptions to only what is absolutely required by the application
  - UDS Packages **may** make use of the [UDS `Exemption` custom resource](https://github.com/defenseunicorns/uds-core/blob/main/src/pepr/operator/README.md#example-uds-exemption-cr) for exempting any Pepr policies, but in doing so they **Must** document rationale for the exemptions
- **Must** declaratively implement any available application hardening guidelines by default (Example: [GitLab Hardening guidelines](https://docs.gitlab.com/ee/security/hardening.html))
- **Must** release a unicorn flavor image, providing a minimal CVE baseline
- **Should** avoid workarounds with Istio such as disabling strict mTLS peer authentication.

> - **Should** minimize network policies to specific selectors needed for Ingress/Egress traffic.

Author: Confirmed they are in silver

- **Should** consider security options during implementation to provide the most secure default possible (i.e. SAML w/SCIM vs OIDC).
- **Should** name the client `<App> Login` (i.e. `Mattermost Login`) to provide login UX consistency.

> - **Should** clearly mark the client id with the group and app name `uds-<group>-<application>` (i.e. `uds-swf-mattermost`) to provide consistency in the Keycloak UI.

Author: Moved to Silver

- **Should** limit the use of Zarf variable templates and prioritize configuring packages via Helm value overrides.
  > This ensures that the package is configured the same way that the bundle would be and avoids any side effect issues of Zarf's `###` templating.

## Silver

_a Silver UDS-Package that integrates with the main features of the UDS Operator, is documented, maintained, and can be confidently operated in production._

- **Must** satisfy all the requirements of [Bronze](#bronze) Packages
- **Must** define network policies under the `allow` key as required in the [UDS Package Custom Resource](https://github.com/defenseunicorns/uds-core/blob/main/docs/configuration/uds-operator.md)

> - **Must** use and create a Keycloak client through the `sso` key if the application provides an end-user login. [UDS Package Custom Resource](https://github.com/defenseunicorns/uds-core/blob/main/docs/configuration/uds-operator.md)

Reviewer 3: ```suggestion
- **Must** (except if the application provides no end user login) use and create a Keycloak client through the `sso` key. [UDS Package Custom Resource](https://github.com/defenseunicorns/uds-core/blob/main/docs/configuration/uds-operator.md)
```
Do we want to call out exceptions in a more standard way - there will be apps (especially stuff like DBs) that you wouldn't create a KC client for and it may be good to callout exceptions up front.

- **Must** define any external interfaces under the `expose` key in the [UDS Package Custom Resource](https://github.com/defenseunicorns/uds-core/blob/main/docs/configuration/uds-operator.md)
- **Must** deploy and operate successfully with Istio injection enabled in the namespace.
- **Must** implement Journey testing, covering the basic user flows and features of the application (see [Testing Guidelines](./testing-guidelines.md))
  > This can by something like Playwright / [Cypress](https://github.com/defenseunicorns/uds-identity-config/tree/main/src/test/cypress) tests for services with a Web UI or something like [Jest](https://github.com/defenseunicorns/uds-package-gitlab-runner/tree/main/test) tests for headless services.
- **Must** implement Upgrade Testing to ensure that the current development package works when deployed over the previously released one. (see [Testing Guidelines](./testing-guidelines.md))
- **Must** be capable of operating within an internet-disconnected (air-gapped) environment
- **Must** be actively maintained by the package maintainers identified in CODEOWNERS
- **Must** be versioned using the UDS Package [Versioning scheme](#versioning)

> - **Must** contain documentation under a `docs` folder at the root that describes how to configure the package and outlines package dependencies.

Reviewer 3: Maybe just my OCD but do we want to still have sub-categories under the levels (like `UDS Core Integrations`, `Package Structure`, and `Maintenance Practices`?)

Author: Yes. Created https://github.com/defenseunicorns/uds-common/issues/255 are you okay defering these changes to that issue? I'm not 100% the best way to organize the information to retain conciseness and provide that level of clarification.

Author: https://github.com/defenseunicorns/uds-common/issues/255

Reviewer 3: I think that's fine to capture in that issue for now.

>   > This allows users of the package to learn more about exposed configuration - it is recommended to make the entrypoint for configuration `configuration.md`.

Author: https://github.com/defenseunicorns/uds-common/issues/247

> - **Must** have a dependency management bot (such as renovate) configured to open PRs to update core package and support dependencies.

Reviewer 3: ```suggestion
- **Must** have a dependency management bot (such as renovate) configured to open PRs to update the core package and support dependencies.
```

- **Must** release its package to the `ghcr.io/defenseunicorns/packages/<group>` namespace as the application's name (i.e. `ghcr.io/defenseunicorns/packages/uds/mattermost`).
- **Must** not make the assumption that the `expose` interfaces are accessible to the bastion or pipeline deploying the package (i.e. `*.uds.dev`). If web requests need to be made they should be done through a `Job` or `./uds zarf tools kubectl exec` as appropriate.
- **Should** lint their configurations with appropriate tooling such as [`yamllint`](https://github.com/adrienverge/yamllint) and [`zarf dev lint`](https://docs.zarf.dev/commands/zarf_dev_lint/).

## Badging

> [!NOTE]
> The badge should link to the `uds-core` repo and should match the height of any other badges (i.e. `20px`).

> [!NOTE]

> > Additional services will be added and integration requirements will change as `uds-core` continues to evolve. UDS Package implementers **should** keep up with `uds-core` releases and changes to this document. The "Made for UDS" badge **may** be removed if a package is not updated in a timely fashion.

Reviewer 3: Do we want to solve this rather than have this statement? May be nice to lay out how this would go (i.e. Package is identified as not meeting standards, maintainer is notified, maintainer has 30 days to fix before their badge is downgraded or removed?)

Author: https://github.com/defenseunicorns/uds-common/issues/256 

I do want to solve this, but think we need to improve CI before we are able to do it effectively.

17.2.1-uds.1
```

## General

And in addition to the above, packages generally:
## Documenting Exceptions

- **Must** be capable of operating within an internet-disconnected (air-gapped) environment.
If you are unable to meet a requirement or best practice, document the justification in your package repository `docs/package_exceptions.md`

- **Must** not make the assumption that the `expose` interfaces are accessible to the bastion or pipeline deploying the package (i.e. `*.uds.dev`).  If web requests need to be made they should be done through a `Job` or `./uds zarf tools kubectl exec` as appropriate.
Some examples of valid justifications would be:
- ✅ The last step of deployment requires a click-ops interaction and cannot be automated through IaC
- ✅ Upstream application does not expose metrics endpoint to integrate

> - **Must** be maintained by a resourced team that is explicitly defined as maintaining the project (i.e. in `CODEOWNERS`).

Reviewer 4: I understand that this is sort of replaced above, but I dont like the implications of (at least for any internal to the company developed package) "resourced team" no longer being in the language.

Author: I added a small section to articulate!

17.2.1-uds.1
```

## Documenting Exceptions

If you are unable to meet a requirement or best practice, document the justification in your package repository `docs/package_exceptions.md`

Some examples of valid justifications would be:

> - ✅ The last step of deployment requires a click-ops interaction and cannot be automated through IaC

Reviewer 3: ```suggestion
- ✅ The last step of SSO configuration requires a click-ops interaction and cannot be automated declaratively
```

Reviewer 3: Not sure if this is actually a valid exception given that this is a **Must** as I am interpreting it - as written the reqs _only_ require integration with the UDS Operator (so SSO, Networking, and Monitoring) so only these configurations **must** be declarative - if something falls outside of these configurations it wouldn't matter and would not need to be documented here (i.e. the final Mattermost setup of a Team for example) - and if an application didn't integrate with one of the UDS Operator features I do think that it would be worth holding it back a level intentionally and if it is important working with upstream to fix it.

> - ✅ Upstream application does not expose metrics endpoint to integrate

Reviewer 3: IMO this one is also odd since the wording `- **Must** implement monitors for each application metrics endpoint using it's built-in chart monitors, "monitor" key, or manual monitors in the config chart.` would not apply if the application doesn't implement monitors - there would be no for each loop to check through - also are we documenting exceptions to Musts or other parts with flexibility (Shoulds/Mays)?  IMO Musts **must** be the things that we will not compromise on and if there are exceptions (i.e. no user login, no metrics endpoint) those should simply be worded into and called out in the Must text.

Author: removed this part :)
